hatch/src/cli/commands/build.rs

```diff
@@ -17,57 +16,65 @@ impl<'build> Build {
+    task::generate_assets(&project).with_context(|e| {
+      format!("Failed to generate assets : {}", e)
+    })?;
+
+    if let Some(path) = project.path().to_str() {
+      let command = format!("cd {} && tup", path);
+      match task::platform_type() {
+        PlatformKind::Windows => {
+          let mut child =
+            process::Command::new("cmd")
+              .arg("/C")
+              .arg(command)
+              .spawn().with_context(|e| {
+              format!("failed to build project at `{}` : {}", &path, e)
+            })?;
+          child.wait()?;
+        },
+        _ => {
+          let mut child =
+            process::Command::new("sh")
+              .arg("-c")
+              .arg(command)
+              .spawn().with_context(|e| {
+              format!("failed to build project at `{}` : {}", &path, e)
+            })?;
+          child.wait()?;
+        }
+      }
+    }
+
+    Ok(())
+  }
 }
 
 impl<'command> Command<'command> for Build {
   fn cli_subcommand(&self) -> App<'command, 'command> {
     SubCommand::with_name(&self.name)
       .about("Builds a project.")
       .author("Josh Gould <mrgould93@gmail.com>")
-
-      .arg(Arg::with_name(PROJECT_NAMES)
-           .help("The projects to be built.")
-           .required(false)
-           .min_values(0)
-           .value_delimiter(" "))
   }
 
   fn subcommand_name(&self) -> &'static str {
     self.name
   }
 
-  fn execute(&self, args: &ArgMatches<'command>) -> HatchResult<Project> {
+  fn execute(&self, args: &ArgMatches<'command>) -> HatchResult<()> {
     let project_path = self.project_path(args);
     let project = task::read_project(&project_path).with_context(|e| {
-      format!("Failed to read project `{}` : {}", project_path.display(), e)
+      format!("Failed to read project at `{}` : {}", project_path.display(), e)
     })?;
 
     task::generate_assets(&project).with_context(|e| {
       format!("Failed to generate assets : {}", e)
     })?;
 
-    if let Some(path) = project.path().to_str() {
-      let command = format!("cd {} && tup", path);
-      match task::platform_type() {
-        PlatformKind::Windows => {
-          let mut child = process::Command::new("cmd")
-                                            .arg("/C")
-                                            .arg(command)
-                                            .spawn()
-                                            .expect("failed to build project.");
-          child.wait();
-        },
-        _ => {
-          let mut child = process::Command::new("sh")
-                                            .arg("-c")
-                                            .arg(command)
-                                            .spawn()
-                                            .expect("failed to build project.");
-          child.wait();
-        }
-      }
-    }
+    println!("Building project...\n");
 
-    Ok(project)
+    self.execute(&project)
```

Comment:
So can we do something like:  
```rust
let res = (|| -> HatchResult<()> {
// implementation of command::execute here
})().with_context(|e| {
  format!("Failed to build project : {}", e)
})?;
Ok(res)
``` 
where we wrap pretty much everything in the implementation of command::execute in a well defined error message with a reason and then make the changes to the error messages I suggested?

Reply:
I am not opposed, I like this exception style error handling pattern :).